src/ai_migrate/s3_uploader.py

```diff
@@ -0,0 +1,60 @@
+from datetime import datetime
+from pathlib import Path
+import boto3
+import asyncio
+import shutil
+import logging
+
+logger = logging.getLogger("ai_migrate.s3_uploader")
+
+
+class S3Uploader:
+    def __init__(self, bucket_name: str):
+        self.s3_client = boto3.client("s3") if bucket_name != "localhost" else None
+        self.bucket = bucket_name
+
+    async def upload(
+        self, project: str, paths: list[Path], result_type: str, timestamp: datetime
+    ) -> None:
+        dest = Path(f"{project}/attempt-{timestamp.strftime('%Y%m%d-%H%M%S')}")
+        if result_type:
+            dest = dest / result_type
+
+        if self.bucket == "localhost":
+            dest = Path("~/ai-migration-results").expanduser() / dest
+            dest.mkdir(parents=True, exist_ok=True)
+            for path in paths:
+                try:
+                    shutil.copy2(path, dest / path.name)
+                except OSError as e:
+                    logger.error(f"Failed to save {path.name}: {e}")
+            return
+
+        async def upload_one(path: Path):
+            key = str(dest / path.name)
+            try:
+                async with asyncio.Semaphore(5):
```

Comment:
If you want to limit the concurrency here, don't you need to create the semaphore outside of the loop/comprehension and enter/exit it?

Reply:
and here I was thinking I was fixing AI code by doing this.... 
100% artisanal human bug...